In [88]:
import pandas as pd
import numpy as np
import requests
import json
import matplotlib.pyplot as plt
%matplotlib inline 
from IPython.display import clear_output

In [89]:
#leitura do arquivo csv 
df = pd.read_csv('twitter-archive-enhanced.csv')


# Avaliação e Arrumação

## 1. Avaliação do arquivo CSV:

* existe muitos elementos na coluna nome nulos, só que preenchidos como NONE.
* a coluna timestamp encontr-ase como string quando deveria ser DATE.

## 2. Arrumação do arquivo CSV:

* acredito que as colunas in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id,   retweeted_status_timestamp são desnecessárias a analise.

In [90]:
df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [92]:
df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [93]:
#verificar dados duplicados
df.duplicated()

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27      False
28      False
29      False
        ...  
2326    False
2327    False
2328    False
2329    False
2330    False
2331    False
2332    False
2333    False
2334    False
2335    False
2336    False
2337    False
2338    False
2339    False
2340    False
2341    False
2342    False
2343    False
2344    False
2345    False
2346    False
2347    False
2348    False
2349    False
2350    False
2351    False
2352    False
2353    False
2354    False
2355    False
Length: 2356, dtype: bool

In [94]:
#verificar dados nulos
df.isnull()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  timestamp  source  \
0        False                   True                 True      False   False   
1        False                   True                 True      False   False   
2        False                   True                 True      False   False   
3        False                   True                 True      False   False   
4        False                   True                 True      False   False   
5        False                   True                 True      False   False   
6        False                   True                 True      False   False   
7        False                   True                 True      False   False   
8        False                   True                 True      False   False   
9        False                   True                 True      False   False   
10       False                   True                 True      False   False   
11       False                   True                 True      False   False   
12       False                   True                 True      False   False   
13       False                   True                 True      False   False   
14       False                   True                 True      False   False   
15       False                   True                 True      False   False   
16       False                   True                 True      False   False   
17       False                   True                 True      False   False   
18       False                   True                 True      False   False   
19       False                   True                 True      False   False   
20       False                   True                 True      False   False   
21       False                   True                 True      False   False   
22       False                   True                 True      False   False   
23       False                   True                 True      False   False   
24       False                   True                 True      False   False   
25       False                   True                 True      False   False   
26       False                   True                 True      False   False   
27       False                   True                 True      False   False   
28       False                   True                 True      False   False   
29       False                   True                 True      False   False   
...        ...                    ...                  ...        ...     ...   
2326     False                   True                 True      False   False   
2327     False                   True                 True      False   False   
2328     False                   True                 True      False   False   
2329     False                   True                 True      False   False   
2330     False                   True                 True      False   False   
2331     False                   True                 True      False   False   
2332     False                   True                 True      False   False   
2333     False                   True                 True      False   False   
2334     False                   True                 True      False   False   
2335     False                   True                 True      False   False   
2336     False                   True                 True      False   False   
2337     False                   True                 True      False   False   
2338     False                   True                 True      False   False   
2339     False                   True                 True      False   False   
2340     False                   True                 True      False   False   
2341     False                   True                 True      False   False   
2342     False                   True                 True      False   False   
2343     False                   True

In [95]:
#verificar a soma dos dados nulos
df.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [96]:
#contagem dos nomes
df['name'].value_counts()

None              745
a                  55
Charlie            12
Oliver             11
Cooper             11
Lucy               11
Lola               10
Penny              10
Tucker             10
Winston             9
Bo                  9
Sadie               8
the                 8
Bailey              7
Daisy               7
an                  7
Buddy               7
Toby                7
Koda                6
Jack                6
Milo                6
Dave                6
Bella               6
Jax                 6
Oscar               6
Rusty               6
Scout               6
Leo                 6
Stanley             6
Louis               5
                 ... 
Eleanor             1
Timmy               1
Miguel              1
Amber               1
Lenox               1
Gin                 1
Lucia               1
Steve               1
Goose               1
Jameson             1
Lilah               1
Mo                  1
Brownie             1
Bayley              1
Randall   

In [97]:
df['source'].value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

## 1. Avaliação do arquivo tsv:

* o arquivo csv mostra 2175 tweets, só que o arquivo tsv nos mostra 2075 imagens dos cães, logo, faltam 100 arquivos de imagens

## 2. Arrumação do arquivo tsv:

* acredito que não se faz necessário três colunas para indicar a raça como p1 que indica nome, p1_conf o código da raça, p1_dog que indica se verdadeiro ou falso. Logo, bastava duas colunas para verificar.

In [98]:
#função para baixar arquivo da página url
def get_file(url):
    name = url.split('/')[-1]
    response = requests.get(url)
    if response.status_code == 200:
        try:
            with open(name, mode='wb') as file:
                file.write(response.content)
            print('Arquivo {} download realizado com sucesso!'.format(name))
        except IOError as err:
            name = ""
            print("Erro de I/O: {}".format(err))
    else:
        name = ""
        print('URL inválida.')
    return name

In [99]:
#baixando arquivo html
image_url= 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
name = get_file(image_url)

Arquivo image-predictions.tsv download realizado com sucesso!


In [100]:
#leitura do aquivo tsv
df_im = pd.read_csv(name, delimiter = '\t')

In [101]:
df_im.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [102]:
df_im.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [103]:
df_im.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [104]:
#contagem dos nomes das raças existentes na tabela
df_im['p1'].value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
malamute                      30
cocker_spaniel                30
French_bulldog                26
Chesapeake_Bay_retriever      23
miniature_pinscher            23
seat_belt                     22
Staffordshire_bullterrier     20
German_shepherd               20
Siberian_husky                20
web_site                      19
Cardigan                      19
Shetland_sheepdog             18
Eskimo_dog                    18
teddy                         18
beagle                        18
Maltese_dog                   18
Rottweiler                    17
Shih-Tzu                      17
Lakeland_terrier              17
kuvasz                        16
Italian_greyhound             16
          

In [105]:
#leitura dos dados json dos tweets a partir API
import tweepy
consumer_key = 'YOUR CONSUMER KEY'
consumer_secret = 'YOUR CONSUMER SECRET'
access_token = 'YOUR ACCESS TOKEN'
access_secret = 'YOUR ACCESS SECRET'

auth = tweepy.OAuthHandler('0ET3LhnJt11VytU68RzKB9yXN', 'dycPKdH5ai6xOQ1QB5EbksDtAXuY3G6n6Jsb34hlxEkMT5kLsP')
auth.set_access_token('1470627588-374awz7oLdw7hwlYBErHAIyIBr5achrOcUkgy7q', '2CCMOkUfGwKZUPCVNLqfCqGiPRqfDpSaVVntT4Q4JgLin')

api = tweepy.API(auth)

In [106]:
#baixando os arquivos json dos tweets e verificação dos erros
tweet_dog = 'tweet_json.txt'

ids = df.tweet_id.values
errors = []

try:
    with open(tweet_dog, 'w') as file:
        for i, id in enumerate(ids):
            clear_output()
            print('Progresso: {:.2f}%\nErros: {}'.format(float((i+1)*100/len(ids)), len(errors)))
            try:
                tweet = api.get_status(id, tweet_mode='extended')
                file.write(json.dumps(tweet._json))
                if i < len(ids) - 1:
                    file.write('\n')
            except tweepy.TweepError:
                errors.append(id)
    clear_output()
    print('Arquivo {} gravado com sucesso!'.format(tweet_dog))
except IOError as err:
    print("Erro de I/O: {}".format(err))

if(len(errors)) > 0:
    error_json_file = 'id_json_errors.txt'
    print('\nErro obtendo id\'s: {}'.format(errors))
    try:
        with open(error_json_file, 'w') as file:
            for id in errors:
                file.write(str(id) + '\n')
            print('\nArquivo {} gravado com sucesso!'.format(error_json_file))
    except IOError as err:
        print("\nErro de I/O: {}".format(err))

Arquivo tweet_json.txt gravado com sucesso!

Erro obtendo id's: [888202515573088257, 873697596434513921, 872668790621863937, 869988702071779329, 866816280283807744, 861769973181624320, 845459076796616705, 842892208864923648, 837012587749474308, 827228250799742977, 812747805718642688, 802247111496568832, 775096608509886464, 770743923962707968, 758740312047005698, 758474966123810816, 758467244762497024, 758405701903519748, 758355060040593408, 758099635764359168, 758041019896193024, 757741869644341248, 757729163776290825, 757725642876129280, 757611664640446465, 757597904299253760, 757596066325864448, 757400162377592832, 757393109802180609, 757354760399941633, 756998049151549440, 756939218950160384, 756651752796094464, 756526248105566208, 756303284449767430, 756288534030475264, 756275833623502848, 755955933503782912, 755206590534418437, 755110668769038337, 754874841593970688, 754856583969079297, 754747087846248448, 754482103782404096, 754449512966619136, 754120377874386944, 754011816964026

In [109]:
#transformando os arquivos json em lista
t_list = []

try:
    with open(tweet_dog) as file:
        for line in file:
            current_json = json.loads(line)
            tweet_id = current_json['id']
            retweet_count = current_json['retweet_count']
            favorite_count = current_json['favorite_count']
            t_list.append({'tweet_id': int(tweet_id),
                        'retweet_count': int(retweet_count),
                        'favorite_count': int(favorite_count)})
except FileNotFoundError as err:                         
    print("\nErro de I/O: {}".format(err))

df_tw = pd.DataFrame(t_list, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

In [110]:
df_tw.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           8290           37952
1  892177421306343426           6122           32592
2  891815181378084864           4054           24536
3  891689557279858688           8428           41293
4  891327558926688256           9130           39483

In [111]:
df_tw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1726 entries, 0 to 1725
Data columns (total 3 columns):
tweet_id          1726 non-null int64
retweet_count     1726 non-null int64
favorite_count    1726 non-null int64
dtypes: int64(3)
memory usage: 40.5 KB


In [112]:
df_tw.describe()

tweet_id  retweet_count  favorite_count
count  1.726000e+03    1726.000000     1726.000000
mean   7.496680e+17    3179.696408     8784.256083
std    7.720814e+16    4920.995385    13044.492768
min    6.660209e+17       0.000000        0.000000
25%    6.740772e+17     495.250000      912.250000
50%    7.609492e+17    1568.500000     3156.000000
75%    8.196806e+17    3872.750000    12012.750000
max    8.924206e+17   60777.000000   140058.000000

# Limpeza

In [113]:
arquivo = df.copy()
image = df_im.copy()
tweet = df_tw.copy()

Definição:

vamos apagar os arquivos que não apresentem imagem, utilizaremos inner join

 Código

In [114]:
#limpar os arquivos id de tweet que não apresentam imagem associados
arquivo = pd.merge(arquivo, tweet, on=['tweet_id'])
arquivo = pd.merge(arquivo, image, on=['tweet_id'])

Teste

In [115]:
arquivo.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1525 entries, 0 to 1524
Data columns (total 30 columns):
tweet_id                      1525 non-null int64
in_reply_to_status_id         14 non-null float64
in_reply_to_user_id           14 non-null float64
timestamp                     1525 non-null object
source                        1525 non-null object
text                          1525 non-null object
retweeted_status_id           70 non-null float64
retweeted_status_user_id      70 non-null float64
retweeted_status_timestamp    70 non-null object
expanded_urls                 1525 non-null object
rating_numerator              1525 non-null int64
rating_denominator            1525 non-null int64
name                          1525 non-null object
doggo                         1525 non-null object
floofer                       1525 non-null object
pupper                        1525 non-null object
puppo                         1525 non-null object
retweet_count                 1525 

Definição:

iremos deletar as colunas aos quais são desnecessárias.

Código

In [116]:
#eleiminar as colunas irrelevantes
arquivo.drop(['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], axis = 1, inplace = True)

Teste

In [117]:
arquivo.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1525 entries, 0 to 1524
Data columns (total 27 columns):
tweet_id                 1525 non-null int64
in_reply_to_status_id    14 non-null float64
in_reply_to_user_id      14 non-null float64
timestamp                1525 non-null object
source                   1525 non-null object
text                     1525 non-null object
expanded_urls            1525 non-null object
rating_numerator         1525 non-null int64
rating_denominator       1525 non-null int64
name                     1525 non-null object
doggo                    1525 non-null object
floofer                  1525 non-null object
pupper                   1525 non-null object
puppo                    1525 non-null object
retweet_count            1525 non-null int64
favorite_count           1525 non-null int64
jpg_url                  1525 non-null object
img_num                  1525 non-null int64
p1                       1525 non-null object
p1_conf                  15

Definição:

trocar a coluna timestamp de string para date

Código

In [118]:
#transformar a coluna timestamp de string em datetime
arquivo.timestamp = pd.to_datetime(arquivo.timestamp)

Teste

In [119]:
arquivo.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1525 entries, 0 to 1524
Data columns (total 27 columns):
tweet_id                 1525 non-null int64
in_reply_to_status_id    14 non-null float64
in_reply_to_user_id      14 non-null float64
timestamp                1525 non-null datetime64[ns]
source                   1525 non-null object
text                     1525 non-null object
expanded_urls            1525 non-null object
rating_numerator         1525 non-null int64
rating_denominator       1525 non-null int64
name                     1525 non-null object
doggo                    1525 non-null object
floofer                  1525 non-null object
pupper                   1525 non-null object
puppo                    1525 non-null object
retweet_count            1525 non-null int64
favorite_count           1525 non-null int64
jpg_url                  1525 non-null object
img_num                  1525 non-null int64
p1                       1525 non-null object
p1_conf            

Definição:

iremos deletar a coluna boolean da tabela image já que apresenta duas colunas para indicar a nome da raça do cachorro, como também o código.

Código

In [120]:
#eliminar as colunas nas tabelas
image.drop(['p1_dog', 'p2_dog', 'p3_dog'], axis = 1, inplace = True)
arquivo.drop(['p1_dog', 'p2_dog', 'p3_dog'], axis = 1, inplace = True)

Teste

In [121]:
image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 9 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p2          2075 non-null object
p2_conf     2075 non-null float64
p3          2075 non-null object
p3_conf     2075 non-null float64
dtypes: float64(3), int64(2), object(4)
memory usage: 146.0+ KB


# Salvar dados limpos

In [122]:
#salvar arquivo limpo
arqlimpo = 'twitter_archive_master.csv'
arquivo.to_csv(arqlimpo, index= False)

In [123]:
df_l = pd.read_csv(arqlimpo)

In [124]:
df_l.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1525 entries, 0 to 1524
Data columns (total 24 columns):
tweet_id                 1525 non-null int64
in_reply_to_status_id    14 non-null float64
in_reply_to_user_id      14 non-null float64
timestamp                1525 non-null object
source                   1525 non-null object
text                     1525 non-null object
expanded_urls            1525 non-null object
rating_numerator         1525 non-null int64
rating_denominator       1525 non-null int64
name                     1525 non-null object
doggo                    1525 non-null object
floofer                  1525 non-null object
pupper                   1525 non-null object
puppo                    1525 non-null object
retweet_count            1525 non-null int64
favorite_count           1525 non-null int64
jpg_url                  1525 non-null object
img_num                  1525 non-null int64
p1                       1525 non-null object
p1_conf                  15